File: server/TWS Foundation/CSM Foundation/Database/Models/Options/Filters/SetViewFilterLinearEvaluation.cs
using System.Linq.Expressions;

using CSM_Foundation.Database.Interfaces;

namespace CSM_Foundation.Database.Models.Options.Filters;
public class SetViewFilterLinearEvaluation<TSet> 
    : ISetViewFilterNode<TSet>
    where TSet : ISet {

    public int Order { get; set; } = 0;
    public string Discrimination { get; } = typeof(SetViewFilterLinearEvaluation<TSet>).ToString();


    public required SetViewFilterEvaluationOperators Operator { get; set; }
    public required ISetViewFilter<TSet>[] Filters { get; set; }

    public Expression<Func<TSet, bool>> Compose() {
        if(Filters.Length <= 0) 
            throw new Exception("Null filters unable to calculate");

        ParameterExpression parameter = Expression.Parameter(typeof(TSet), "x");


        Expression resExpression = Expression.Invoke(Filters[0].Compose(), parameter);
        for (int i = 1; i < Filters.Length; i++) {
            ISetViewFilter<TSet> filter = Filters[i];

            Expression expression = Expression.Invoke(filter.Compose(), parameter);

            switch (Operator) {
                case SetViewFilterEvaluationOperators.OR:
                    resExpression = Expression.OrElse(resExpression, expression);
                    break;
                case SetViewFilterEvaluationOperators.AND:
                    resExpression = Expression.AndAlso(resExpression, expression);
                    break;
            }
        }

        return Expression.Lambda<Func<TSet, bool>>(resExpression, parameter);
    }
}


public enum SetViewFilterEvaluationOperators {
    OR, 
    AND,
}

File: client/tws_foundation_client/lib/src/models/options/filters/set_view_filter_linear_evaluation.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/src/models/interfaces/set_view_filter_interface.dart';
import 'package:tws_foundation_client/src/models/interfaces/set_view_filter_node_interface.dart';

///
final class SetViewFilterLinearEvaluation<TSet extends CSMSetInterface> implements SetViewFilterNodeInterface<TSet> {
  ///
  @override
  final String discrimination = 'SetViewFilterLinearEvaluation`1';

  ///
  @override
  final int order;

  ///
  final SetViewFilterEvaluationOperators operator;

  ///
  final List<SetViewFilterInterface<TSet>> filters;

  ///
  SetViewFilterLinearEvaluation(this.order, this.operator, this.filters);

  @override
  JObject encode() {
    return <String, dynamic>{
      'discrimination': discrimination,
      'order': order,
      'operator': operator.index,
      'filters': filters.map((SetViewFilterInterface<TSet> i) => i.encode()).toList(),
    };
  }
}

///
enum SetViewFilterEvaluationOperators {
  ///
  or,

  ///
  and,
}
